orthanc/orthanc-anon/plugin/pixl.py

```diff
@@ -219,42 +216,151 @@ def OnHeartBeat(output, uri, **request) -> Any:  # noqa: ARG001
+
+    - Pull a study from Orthanc Raw based on its resource ID. Wait for the study to be stable.
+    - Iterate over instances and anonymise them
+    - Re-upload the study via the dicom-web api. Wait for the study to be stable.
+    - Notify the PIXL export-api to send the study the to relevant endpoint
+    """
+    payload = json.loads(request["body"])
+    study_uid = payload["StudyInstanceUID"]
+    query_id = payload["QueryID"]
+    retrieve_response = json.loads(
+        orthanc.RestApiPost(f"/queries/{query_id}/retrieve", json.dumps({}))
+    )
+    logger.debug(
+        "Response from retrieving study {} from Orthan Raw: {}", study_uid, retrieve_response
+    )
+
+    # Download the zipped study from Orthanc Anon
+    study_resource_id = _get_existing_study(study_uid=study_uid)
+    if study_resource_id is None:
+        return
+    zipped_study_bytes = BytesIO(orthanc.RestApiGet(f"/studies/{study_resource_id}/archive"))
+    logger.trace("Study data response {}", zipped_study_bytes)
+
+    # Delete the oringal study now in case anything goes wrong with the anonymisation.
+    # We don't want to leave the original (non-anonymised) study on Orthanc Anon
+    logger.info(
+        "Deleteing non-anonymised study with UID {} and resource ID {} from Orthanc Anon",
+        study_uid,
+        study_resource_id,
+    )
+    orthanc.RestApiDelete(f"/studies/{study_resource_id}")
+
+    # Anonymise the study, re-upload to Orthanc Anon, and notify the export API to export the study
+    with ZipFile(zipped_study_bytes) as zipped_study:
+        anonymised_instances_bytes, anonymised_study_uid = _anonymise_study_instances(
+            zipped_study=zipped_study,
+            study_uid=study_uid,
+        )
+    _upload_instances(anonymised_instances_bytes)
+    anonymised_study_resource_id = _get_existing_study(study_uid=anonymised_study_uid)
+    Send(study_id=anonymised_study_resource_id)
+
+
+def _get_existing_study(study_uid: str) -> Optional[str]:
+    """
+    Get the resource ID for an existing study based on its StudyInstanceUID.
+
+    Returns None if there are no resources with the given StudyInstanceUID.
+    Returns the resource ID if there is a single resource with the given StudyInstanceUID.
+    Returns None if there are multiple resources with the given StudyInstanceUID and deletes
+    the studies.
+    """
+    data = json.dumps(
+        {
+            "Level": "Study",
+            "Query": {
+                "StudyInstanceUID": study_uid,
+            },
+        }
+    )
+    study_resource_ids = json.loads(orthanc.RestApiPost("/tools/find", data))
+    if not study_resource_ids:
+        logger.info(f"No study found with StudyInstanceUID {study_uid}")
+        return None
+
+    if len(study_resource_ids) == 1:
+        return study_resource_ids[0]
+
+    logger.info(
+        f"{len(study_resource_ids)} found with StudyInstanceUID {study_uid}. Deleting studies."
+    )
+
+    return None
```

Comment:
is this check needed? Is it possible that multiple studies will exist?

Also, do we need to check the project name is as expected? Is it possible that the same study is being exported by two projects, and the imaging api modifies the project name before Orthanc Anon retrieves the data?

Reply:
I don't think it would be possible that multiple studies exist. Is possible but feels very unlikely that two studies would be anonymised at the same time, and if they are the winning one will work and the second one will just need to be re-published again so probably fine

Reply:
oh okay, that simplifies things then